benchmarks/python/optimized-numpy.py

```diff
@@ -0,0 +1,123 @@
+import sys
+import math
+import timeit
+
+import numpy as np
+import pandas as pd
+
+def load_input_data(path):
+
+    df = pd.read_csv(
+        path, names = ["mass", "x", "y", "z", "vx", "vy", "vz"], delimiter = r"\s+"
+    )
+
+    masses = df["mass"].values.copy()
+    positions = df.loc[:, ["x", "y", "z"]].values.copy()
+    velocities = df.loc[:, ["vx", "vy", "vz"]].values.copy()
+
+    return masses, positions, velocities
+
+def compute_accelerations(accelerations, masses, positions):
+    nb_particles = masses.size
+
+    for index_p0 in range(nb_particles - 1):
+        position0 = positions[index_p0]
+        mass0 = masses[index_p0]
+
+        vectors = position0 - positions[index_p0 + 1: nb_particles]
+
+        distances = np.square(vectors).sum(axis = 1)
+        coefs = distances ** 1.5
+
+        _x = np.add(np.sum(
+            np.divide(
+                np.multiply(
+                    masses[index_p0 + 1: nb_particles], -1 * vectors.T
+                    ),
+                coefs).T, axis = 0),
+            accelerations[index_p0]
+            )
+        accelerations[index_p0] = _x
+
+        _temp = np.add(
+                np.divide(
+                np.multiply(
+                    mass0, vectors.T
+                    ),
+                coefs).T,
+            accelerations[index_p0 + 1: nb_particles]
+            )
+        accelerations[index_p0 + 1: nb_particles] = _temp
+
+    return accelerations
+
+def numpy_loop(
+        time_step: float,
+        nb_steps: int,
+        masses: "float[]",
+        positions: "float[:,:]",
+        velocities: "float[:,:]",
+    ):
+
+    accelerations = np.zeros_like(positions)
+    accelerations1 = np.zeros_like(positions)
+
+    accelerations = compute_accelerations(accelerations, masses, positions)
+
+    time = 0.0
+
+    energy0, _, _ = compute_energies(masses, positions, velocities)
+    energy_previous = energy0
+
+    for step in range(nb_steps):
+        positions = sum(np.multiply(velocities, time_step), 0.5 * np.multiply(accelerations, time_step ** 2)) + positions
+
+        accelerations, accelerations1 = accelerations1, accelerations
+        accelerations.fill(0)
+        accelerations = compute_accelerations(accelerations, masses, positions)
+
+        velocities = 0.5 * np.multiply(time_step, np.add(accelerations, accelerations1)) + velocities
```

Comment:
In this case, we are multiplying a scaler `time_step` with an array `np.add(accelerations, accelerations1)`. I created a test case to check their behavior and got the following output. The results are absurd. We are getting different results while running the same file multiple times. 

I implemented `np.multiply()` because I felt NumPy's code should consist more of NumPy functions than Python's operator. This might not be a strong reason or a valid reason. Would like to know your viewpoints?

|  | np.multiply() | scaler * array      |
| :---         |     :---:      |          :---: |
 | 100         | 77.6910         | 70.4279      |
 | 10000    | 6722.0320     | 5974.8909   |
 | 100000  | 59978.6810  | 59116.9250  |
 | 100000  | 58738.2830  | 59404.4569  |
 | 100000  | 60946.5689  | 58564.1380  |

Reply:
Did you try reversing the order of your benchmarking? When we tried this a few weeks ago, the results were that the two patterns take the same time, plus minus noise and measuring variations. I think we should try to use python operators as much as possible, it makes the code much more readable and conveys the message that NumPy performance can be achieved without needing special function calls.

Reply:
Yes, I did implement and observed that the time taken by the program to run is reversed. 
Thanks for specifying the reason! I will change NumPy's functions to Python's functions.

Reply:
This seems fixed now, I'll resolve it. If there is still anything that needs preserving for later, please unresolve.